third_party/com.google.cloud.tools.eclipse.jst.server.core/src/com/google/cloud/tools/eclipse/jst/server/core/BasePublishOperation.java

```diff
@@ -0,0 +1,359 @@
+      // avoid changes if no changes to module since last publish
+      IModuleResourceDelta[] delta = getPublishedResourceDelta(module);
+      if (delta == null || delta.length == 0)
+        return;
+    }
+
+    // make directory if it doesn't exist
+    if (!path.toFile().exists())
+      path.toFile().mkdirs();
+
+    IModuleResource[] mr = getResources(module);
+    IStatus[] stat = helper.publishZip(mr, jarPath, monitor);
+    addArrayToList(status, stat);
+    p.put(module[1].getId(), jarURI);
+  }
+
+  private void publishArchiveModule(String jarURI, Properties p, List<IStatus> status,
+      IProgressMonitor monitor) {
+    IPath path = getModuleDeployDirectory(module[0]);
+    boolean moving = false;
+    // Get URI used for previous publish, if known
+    String oldURI = (String) p.get(module[1].getId());
+    if (oldURI != null) {
+      // If old URI found, detect if jar is moving or changing its name
+      if (jarURI != null) {
+        moving = !oldURI.equals(jarURI);
+      }
+    }
+    // If we don't have a jar URI, make a guess so we have one if we need it
+    if (jarURI == null) {
+      jarURI = "WEB-INF/lib/" + module[1].getName();
+    }
+    IPath jarPath = path.append(jarURI);
+    // Make our best determination of the path to the old jar
+    IPath oldJarPath = jarPath;
+    if (oldURI != null) {
+      oldJarPath = path.append(oldURI);
+    }
+    // Establish the destination directory
+    path = jarPath.removeLastSegments(1);
+
+    // Remove if requested or if previously published and are now serving without publishing
+    if (moving || kind == IServer.PUBLISH_CLEAN || deltaKind == ServerBehaviourDelegate.REMOVED
+        || isServeModulesWithoutPublish()) {
+      File file = oldJarPath.toFile();
+      if (file.exists()) {
+        file.delete();
+      }
+      p.remove(module[1].getId());
+
+      if (deltaKind == ServerBehaviourDelegate.REMOVED
+          || isServeModulesWithoutPublish())
+        return;
+    }
+    if (!moving && kind != IServer.PUBLISH_CLEAN && kind != IServer.PUBLISH_FULL) {
+      // avoid changes if no changes to module since last publish
+      IModuleResourceDelta[] delta = getPublishedResourceDelta(module);
+      if (delta == null || delta.length == 0)
+        return;
+    }
+
+    // make directory if it doesn't exist
+    if (!path.toFile().exists())
+      path.toFile().mkdirs();
+
+    IModuleResource[] mr = getResources(module);
+    IStatus[] stat = helper.publishToPath(mr, jarPath, monitor);
+    addArrayToList(status, stat);
+    p.put(module[1].getId(), jarURI);
+  }
+
+  /**
+   * Save the locations of the configured modules.
+   */
+  protected void saveModulePublishLocations(Properties p) {
+    IPath path = getRuntimeBaseDirectory().append("publish.txt");
+    try (FileOutputStream fout = new FileOutputStream(path.toFile())) {
+      p.store(fout, "Module publish data");
+    } catch (Exception e) {
```

Comment:
Do you really need to catch Exception? I would expect only IOException. And why do we ignore this?